src/main/java/dev/codesupport/web/api/service/ArticleServiceImpl.java

```diff
@@ -0,0 +1,225 @@
+        ) {
+            this.articleRepository = articleRepository;
+            this.tagRepository = tagRepository;
+            this.tagSetRepository = tagSetRepository;
+            this.tagSetToTagsRepository = tagSetToTagsRepository;
+        }
+
+        Consumer<ArticleEntity> preSaveLogic() {
+            return (ArticleEntity articleEntity) -> {
+                updateTagReferences(articleEntity);
+                updateArticleMetaData(articleEntity);
+            };
+        }
+
+        Consumer<ArticleEntity> preGetLogic() {
+            return (ArticleEntity articleEntity) -> {
+                Optional<TagSetEntity> optional = tagSetRepository.findById(articleEntity.getTagSetId());
+                if (optional.isPresent()) {
+                    TagSetEntity tagSet = optional.get();
+                    List<TagSetToTagEntity> xMaps = tagSetToTagsRepository.findAllByTagSetIdOrderByTagId(tagSet.getId());
+                    List<TagEntity> tags = xMaps.stream().map(TagSetToTagEntity::getTag).collect(Collectors.toList());
+                    tagSet.setTags(tags);
+                    articleEntity.setTagSet(tagSet);
+                } else {
+                    log.error("Article (ID: " + articleEntity.getId() + ") had no valid TagSet (ID: " + articleEntity.getTagSetId() + ")");
+                }
+            };
+        }
+
+        void updateArticleMetaData(ArticleEntity articleEntity) {
+            Long articleId = articleEntity.getId();
+            if (articleId != null && articleId != 0) {
+                updateExistingArticle(articleEntity);
+            } else {
+                updateNewArticle(articleEntity);
+            }
+        }
+
+        void updateExistingArticle(ArticleEntity articleEntity) {
+            Optional<ArticleEntity> optional = articleRepository.findById(articleEntity.getId());
+
+            if (optional.isPresent()) {
+                ArticleEntity existingArticle = optional.get();
+                articleEntity.setArticleCode(existingArticle.getArticleCode());
+
+                // If the original was finalized, this is a new entry, to preserve history
+                if (existingArticle.isFinalized()) {
+                    articleEntity.setId(null);
+                }
+
+                articleEntity.setUpdatedOn(System.currentTimeMillis());
+            } else {
+                throw new ResourceNotFoundException(ResourceNotFoundException.Reason.NOT_FOUND);
+            }
+        }
+
+        void updateNewArticle(ArticleEntity articleEntity) {
+            // Create an ArticleId which is used to relate various versions of the same article
+            articleEntity.setArticleCode(RandomStringUtils.randomAlphabetic(50));
+            long time = System.currentTimeMillis();
+            articleEntity.setCreatedOn(time);
+            articleEntity.setUpdatedOn(time);
+        }
+
+        void updateTagReferences(ArticleEntity articleEntity) {
```

Comment:
Codacy found an issue: [Use explicit scoping instead of the default package private level](https://app.codacy.com/gh/codesupport/website-backend/pullRequest?prid=6300789)